src/Extension/ChallengeGraphProgressionService.cs

```diff
@@ -0,0 +1,155 @@
+﻿using Extension.Utilities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Extension
+{
+    public class ChallengeGraphProgressionService
+    {
+        public OrderedTable<Challenge> Challenges { get; private set; } = new();
+        public Dictionary<Challenge, List<Challenge>> Dependencies { get; private set; } = new();
+        public Dictionary<Challenge, List<Challenge>> Dependents { get; private set; } = new();
+        public Dictionary<Challenge, bool> HasPassed { get; private set; } = new();
+        public Challenge CurrentChallenge => _lesson.CurrentChallenge;
+
+        private Lesson _lesson = new();
+
+        public ChallengeGraphProgressionService(Lesson lesson)
+        {
+            _lesson = lesson;
+        }
+
+        public ChallengeGraphProgressionService(Lesson lesson, IReadOnlyList<Challenge> challenges)
+            : this(lesson)
+        {
+            foreach (var challenge in challenges)
+            {
+                Challenges.Add(challenge);
+            }
+        }
+
+        public void AddChallenge(Challenge challenge, bool hasPassed = false)
+        {
+            Challenges.Add(challenge);
+            HasPassed.Add(challenge, hasPassed);
+        }
+
+        public void Pass()
+        {
+            Pass(CurrentChallenge);
+        }
+
+        public void Pass(Challenge challenge)
+        {
+            HasPassed[challenge] = true;
+        }
+
+        public async Task<bool> TryGoToNextChallengeAsync()
+        {
+            var firstDependentWithResolvedDependencies = Dependents[CurrentChallenge]
+                .FirstOrDefault(dependent => Dependencies[dependent].All(dependency => HasPassed[dependency]));
+            if (firstDependentWithResolvedDependencies is null)
+            {
+                return false;
+            }
+            else
+            {
+                await _lesson.StartChallengeAsync(firstDependentWithResolvedDependencies);
+                return true;
+            }
+        }
+
+        public async Task<bool> TryGoToFurthestChallengeAsync()
+        {
+            // try to go to the "most advanced" challenges in the graph
+            // "most advanced" challenges
+            // means the set of challenges who is the furthest from the current challenge
+            // in the transpose of the dependency graph (which is dependent graph)
+            // who has all dependency challenges passed
+            // use BFS in the dependent graph
+            await Task.Run(() => { });
```

Comment:
You can also use `await Task.Yield();` if you want to make sure the work doesn't start on the calling thread.